### ISSUE TITLE
Not comprehensive scan

### ISSUE THREAD

**Issue by jhadjar** [rockymeza/wifi#37]
TL;DR: I have a more comprehensive and exhaustive scan with iwlist wlan0 scan than using `wifi.Cell.all('wlan0')`.

The point is: This is random. Sometimes, the module shows all available networks. Sometimes it only shows the one I'm connected to.

Background:

I have a router and a repeater. Both "broadcast" the same SSID, but obviously have different MAC addresses.

I need to change the Wifi password through the router interface (telnet). What this does is that the repeater can't access the router anymore, but still serves the SSID and has the strongest signal, so I need to connect to the router _specifically_.

When I do a `iwlist wlan0 scan` , I _always_ have the results.

When I use the module, it depends. So I have to stop the Network Manager and start it again, wait 10 seconds and do a scan to get the networks' list.

Additional info for memo:

Router: SpeedTouch 585v6.
Repeater: Dlink DIR-505L.

**Comment by Develoman** [comment]
Hi,
I get also the same Bug:
``` python
Traceback (most recent call last):
  File "/home/.../workspace/client-wifi-configuration/wifimanage/linux/__init__.py", line 39, in <module>
    scheme.activate()
  File "/home/.../workspace/client-wifi-configuration/venv/local/lib/python2.7/site-packages/wifi/scheme.py", line 155, in activate
    subprocess.check_output(['/sbin/ifdown', self.interface], stderr=subprocess.STDOUT)
  File "/usr/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['/sbin/ifdown', 'wlan0']' returned non-zero exit status 1
```
and the code:
``` python
    cell  =  Cell.all('wlan0')[ 0 ]
    scheme  =  Scheme.for_cell('wlan0','home', cell,'pw')
    scheme.save() #success save scheme in /etc/network/interfaces
    scheme.activate()
```
the scheme : 
``` python
iface wlan0-home inet dhcp
    wpa-psk 2252012b5441f871ad6e27192d0a9d81797fded0be25525d9fb2c9e86333c772
    wpa-ssid ....
    wireless-channel auto
```

**Comment by theja2289** [comment]
Looks like I'm facing the same issue, I wonder if there is a nice solution to this rather than the work around, I can't afford to stop my network service at any instant as I have a computer which has ssh'd and would keep constantly communicating. I have tried using iwlist as well and I did put my file in the sudoers file. But the iwlist command fails randomly when i call it through subprocess.check_output.

**Comment by rikoz** [comment]
Am getting same bug:
```
Traceback (most recent call last):
  File "/home/.../workspace/client-wifi-configuration/wifimanage/linux/__init__.py", line 39, in <module>
    scheme.activate()
  File "/home/.../workspace/client-wifi-configuration/venv/local/lib/python2.7/site-packages/wifi/scheme.py", line 155, in activate
    subprocess.check_output(['/sbin/ifdown', self.interface], stderr=subprocess.STDOUT)
  File "/usr/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command '['/sbin/ifdown', 'wlan0']' returned non-zero exit status 1
```

**Comment by rikoz** [comment]
From what I was able to find from ifconfig command, my wi-fi network interface is wlp2s0 as to wlan0... Please ther should be a way of auto-detecting the network interface and also in this code:
`scheme = Scheme.for_cell('wlan0', 'home', cell, passkey)`
what is the 'home' argument doing and what type(str, int, bool....) is passed into cell and passkey?

**Comment by melvyn-sopacua** [comment]
@rikoz 
Could you please try:

    pip install wifi==0.8.0rc1

Line 155 referenced doesn't contain that code and I've looked at the current scheme.py and there are no more hardcoded references to wlan0.

**Comment by rikoz** [comment]
@melvyn-sopacua done that

**Comment by rikoz** [comment]
is everything working on version 0.8.0rc1??, perhaps i would like us to have a video chat on this

**Comment by AleRiccardi** [comment]
Hi everyone, I have your same problem, did you fix with `pip install wifi==0.8.0rc1` or did you find other way?

**Comment by rikoz** [comment]
yes, a bit fixed and had another way of auto-detect my wifi interface

**Comment by AleRiccardi** [comment]
Ok because for me didn't work, I will try to find a way and share it here if I success.

**Comment by rikoz** [comment]
![screenshot from 2018-05-02 17-12-58](https://user-images.githubusercontent.com/6421629/39535650-5e092974-4e2c-11e8-9e56-7ebcd7182bab.png)
first detect interface name
.....
then use here: `
# -*- coding: utf-8 -*-

import wifi
from sik_wireless import sik_iface

iface_name = sik_iface()

def Search():
    wifilist = []

    cells = wifi.Cell.all(iface_name)

    for cell in cells:
        wifilist.append(cell)

    return wifilist


def FindFromSearchList(ssid):
    wifilist = Search()

    for cell in wifilist:
        if cell.ssid == ssid:
            return cell

    return False


def FindFromSavedList(ssid):
    cell = wifi.Scheme.find(iface_name, ssid)

    if cell:
        return cell

    return False


def Connect(ssid, password=None):
    cell = FindFromSearchList(ssid)

    if cell:
        savedcell = FindFromSavedList(cell.ssid)

        # Already Saved from Setting
        if savedcell:
            savedcell.activate()
            return cell

        # First time to conenct
        else:
            if cell.encrypted:
                if password:
                    scheme = Add(cell, password)

                    try:
                        scheme.activate()

                    # Wrong Password
                    except wifi.exceptions.ConnectionError:
                        Delete(ssid)
                        return False

                    return cell
                else:
                    return False
            else:
                scheme = Add(cell)

                try:
                    scheme.activate()
                except wifi.exceptions.ConnectionError:
                    Delete(ssid)
                    return False

                return cell
    
    return False


def Add(ssid, password=None):
	cell = FindFromSearchList(ssid)
	
	if not cell:
		return False

	scheme = wifi.Scheme.for_cell(iface_name, ssid, cell, password)
	scheme.save()
	return scheme


def Delete(ssid):
    if not ssid:
        return False

    cell = FindFromSavedList(ssid)

    if cell:
        cell.delete()
        return True

    return False


if __name__ == '__main__':
    # Search wifi and return wifi list
    print Search()

    # Connect wifi with password & without password
    print Connect('Openwifi')
    print FindFromSearchList('koz..py')
    print Add('#koz iPhone', 'omocc5n4jr7zn')
    print FindFromSavedList('#koz iPhone')
    print Connect('#koz iPhone', 'omocc5n4jr7zn')

    # Delete wifi from auto connect list
    #print Delete('Deletewifi')`

**Comment by AleRiccardi** [comment]
rikoz for what I understood I cannot use the `Scheme.activate()` because it produce the error reported  in this conversation and also searching online I saw that it's something related with that shell command `sudo ifup wlan0` that respond me with this messagge `ifup: unknown interface wlan0`.

**Comment by AleRiccardi** [comment]
I found the solution for my problem that is not related to the real question (scan network), but I want to share it. 
In my case I'm not working with libraries but directly with the shell, here I'm passing the SSID and PSW to the `wifi_connect(ssid, psk)` that it return the ip address of my device.

SO: _rasbian stretch_

    import os
    import time
    import subprocess
    
    wpa_supplicant_conf = "/etc/wpa_supplicant/wpa_supplicant.conf"
    sudo_mode = "sudo "
    
    def wifi_connect(ssid, psk):
        cmd_result = ""
    
        # write wifi config to file
        f = open('wifi.conf', 'w')
        f.write('ctrl_interface=DIR=/var/run/wpa_supplicant GROUP=netdev\n')
        f.write('update_config=1\n')
        f.write('country=US\n')
        f.write('\n')
        f.write('network={\n')
        f.write('    ssid="' + ssid + '"\n')
        f.write('    psk="' + psk + '"\n')
        f.write('    key_mgmt=WPA-PSK\n')
        f.write('}\n')
        f.close()
        time.sleep(1)
    
        # move to the specific folder and overwrite the old file
        cmd = 'sudo mv wifi.conf ' + wpa_supplicant_conf
        cmd_result = os.system(cmd)
        print(cmd + " - " + str(cmd_result))
        time.sleep(1)
    
        # reconfigure the wifi service
        cmd = 'wpa_cli -i wlan0 reconfigure'
        cmd_result = os.system(cmd)
        print(cmd + " - " + str(cmd_result))
        time.sleep(10)
    
        # keep all the information about the network
        p = subprocess.Popen(['ifconfig', 'wlan0'], stdout=subprocess.PIPE,
                             stderr=subprocess.PIPE)
    
        out, err = p.communicate()
        ip_address = "<Not Set>"
    
        # extract the IP address
        for l in out.split(b'\n'):
            if l.strip().startswith(b'inet '):
                ip_address = l.strip().split(b'inet ')[1].split(b' ')[0]
    
        return ip_address
    
    
    ip = wifi_connect('SSID', 'PSW')
    
    if ip == "<Not Set>":
        print('Fail to connect to Internet')
    else:
        print('Connection established: IP address: ' + ip.decode("utf-8"))
